File: src/codegen/prelude.py
# SpecSpec Python Prelude
# Validation primitives - embedded at top of generated validators

from typing import Any, Callable, TypeAlias
import os
import json
import zipfile
import re

# Type aliases
Issues: TypeAlias = list[dict[str, Any]]
Validator: TypeAlias = Callable[[Any, list[str], Issues], None]


def add_issue(issues: Issues, path: list[str], code: str, message: str) -> None:
    """Add a validation issue."""
    issues.append({"path": ".".join(path) if path else "(root)", "code": code, "message": message})


# === Primitive validators ===

def validate_str(value: Any, path: list[str], issues: Issues,
                 min_length: int | None = None,
                 max_length: int | None = None,
                 pattern: str | None = None) -> None:
    """Validate string value."""
    if not isinstance(value, str):
        add_issue(issues, path, "type.mismatch", f"Expected string, got {type(value).__name__}")
        return
    if min_length is not None and len(value) < min_length:
        add_issue(issues, path, "str.too_short", f"String length {len(value)} is less than minimum {min_length}")
    if max_length is not None and len(value) > max_length:
        add_issue(issues, path, "str.too_long", f"String length {len(value)} exceeds maximum {max_length}")
    if pattern is not None:
        if not re.match(pattern, value):
            add_issue(issues, path, "str.pattern_mismatch", f"String does not match pattern {pattern}")


def validate_num(value: Any, path: list[str], issues: Issues,
                 min_val: float | None = None,
                 max_val: float | None = None,
                 integer: bool = False) -> None:
    """Validate number value."""
    if not isinstance(value, (int, float)) or isinstance(value, bool):
        add_issue(issues, path, "type.mismatch", f"Expected number, got {type(value).__name__}")
        return
    if integer and not isinstance(value, int):
        add_issue(issues, path, "num.not_integer", f"Expected integer, got {value}")
    if min_val is not None and value < min_val:
        add_issue(issues, path, "num.too_small", f"Number {value} is less than minimum {min_val}")
    if max_val is not None and value > max_val:
        add_issue(issues, path, "num.too_large", f"Number {value} exceeds maximum {max_val}")


def validate_bool(value: Any, path: list[str], issues: Issues) -> None:
    """Validate boolean value."""
    if not isinstance(value, bool):
        add_issue(issues, path, "type.mismatch", f"Expected boolean, got {type(value).__name__}")


def validate_literal(value: Any, path: list[str], issues: Issues, expected: Any) -> None:
    """Validate literal value."""
    if value != expected:
        add_issue(issues, path, "literal.mismatch", f"Expected {expected!r}, got {value!r}")


def validate_pattern(value: Any, path: list[str], issues: Issues, pattern: str) -> None:
    """Validate value matches regex pattern."""
    if not isinstance(value, str):
        add_issue(issues, path, "type.mismatch", f"Expected string for pattern match, got {type(value).__name__}")
        return
    if not re.match(pattern, value):
        add_issue(issues, path, "pattern.mismatch", f"Value does not match pattern {pattern}")


# === Structural validators ===

def validate_object(value: Any, path: list[str], issues: Issues) -> bool:
    """Check value is an object (dict). Returns False if not."""
    if not isinstance(value, dict):
        add_issue(issues, path, "type.mismatch", f"Expected object, got {type(value).__name__}")
        return False
    return True


def validate_field(obj: Any, path: list[str], issues: Issues,
                   key: str, validator: Validator | None = None,
                   optional: bool = False) -> None:
    """Validate a field in an object."""
    if not isinstance(obj, dict):
        return  # Parent validation will catch this

    if key not in obj:
        if not optional:
            add_issue(issues, path, "field.missing", f"Missing required field: {key}")
        return

    if validator is not None:
        validator(obj[key], path + [key], issues)


def validate_list(value: Any, path: list[str], issues: Issues,
                  item_validator: Validator | None = None,
                  min_items: int | None = None,
                  max_items: int | None = None) -> None:
    """Validate list/array value."""
    if not isinstance(value, list):
        add_issue(issues, path, "type.mismatch", f"Expected array, got {type(value).__name__}")
        return

    if min_items is not None and len(value) < min_items:
        add_issue(issues, path, "list.too_short", f"Array length {len(value)} is less than minimum {min_items}")
    if max_items is not None and len(value) > max_items:
        add_issue(issues, path, "list.too_long", f"Array length {len(value)} exceeds maximum {max_items}")

    if item_validator is not None:
        for i, item in enumerate(value):
            item_validator(item, path + [f"[{i}]"], issues)


def validate_oneof(value: Any, path: list[str], issues: Issues,
                   validators: list[Validator],
                   descriptions: list[str] | None = None) -> None:
    """Validate value matches one of the validators."""
    for validator in validators:
        test_issues: Issues = []
        validator(value, path, test_issues)
        if not test_issues:
            return  # Matched

    desc = ", ".join(descriptions) if descriptions else "any of the options"
    add_issue(issues, path, "oneof.no_match", f"Value does not match {desc}")


def matches_validator(value: Any, validator: Validator) -> bool:
    """Test if value matches a validator without adding issues."""
    test_issues: Issues = []
    validator(value, [], test_issues)
    return len(test_issues) == 0


# === File system context ===

class FSContext:
    """File system access context - abstracts directory and zip file access."""

    def __init__(self, base_path: str):
        self.base_path = base_path
        self.is_zip = False
        self.zip_file: zipfile.ZipFile | None = None

        if os.path.isdir(base_path):
            self.is_zip = False
        elif os.path.isfile(base_path) and zipfile.is_zipfile(base_path):
            self.is_zip = True
            self.zip_file = zipfile.ZipFile(base_path, 'r')
        else:
            raise ValueError(f"Not a directory or zip file: {base_path}")

    def exists(self, rel_path: str) -> bool:
        """Check if a file or directory exists."""
        if self.is_zip:
            try:
                self.zip_file.getinfo(rel_path)
                return True
            except KeyError:
                # Try with trailing slash for directories
                try:
                    self.zip_file.getinfo(rel_path + '/')
                    return True
                except KeyError:
                    return False
        else:
            return os.path.exists(os.path.join(self.base_path, rel_path))

    def is_file(self, rel_path: str) -> bool:
        """Check if path is a file."""
        if self.is_zip:
            try:
                info = self.zip_file.getinfo(rel_path)
                return not info.is_dir()
            except KeyError:
                return False
        else:
            return os.path.isfile(os.path.join(self.base_path, rel_path))

    def is_dir(self, rel_path: str) -> bool:
        """Check if path is a directory."""
        if self.is_zip:
            try:
                info = self.zip_file.getinfo(rel_path + '/')
                return info.is_dir()
            except KeyError:
                # Check if any files start with this path
                prefix = rel_path + '/'
                return any(name.startswith(prefix) for name in self.zip_file.namelist())
        else:
            return os.path.isdir(os.path.join(self.base_path, rel_path))

    def read(self, rel_path: str) -> str:
        """Read file content as string."""
        if self.is_zip:
            return self.zip_file.read(rel_path).decode('utf-8')
        else:
            with open(os.path.join(self.base_path, rel_path), 'r', encoding='utf-8') as f:
                return f.read()

    def read_json(self, rel_path: str) -> Any:
        """Read and parse JSON file."""
        return json.loads(self.read(rel_path))

    def basename(self) -> str:
        """Get the base name of the bundle (without extension)."""
        name = os.path.basename(self.base_path)
        if '.' in name:
            name = name.rsplit('.', 1)[0]
        return name

    def close(self) -> None:
        """Close resources."""
        if self.zip_file:
            self.zip_file.close()


# === File system validators ===

FSValidator: TypeAlias = Callable[['FSContext', list[str], Issues], None]


def validate_bundle(path: str, path_list: list[str], issues: Issues,
                    accept_dir: bool = True,
                    accept_zip: bool = False,
                    zip_ext: str | None = None,
                    name_pattern: str | None = None,
                    content_validator: FSValidator | None = None) -> FSContext | None:
    """
    Validate a bundle (directory or zip file).
    Returns FSContext if successful, None otherwise.
    """
    is_dir = os.path.isdir(path)
    is_zip = not is_dir and os.path.isfile(path) and zipfile.is_zipfile(path)

    # Check accepted types
    if is_dir and not accept_dir:
        add_issue(issues, path_list, "bundle.type_mismatch", "Directory not accepted")
        return None
    if is_zip and not accept_zip:
        add_issue(issues, path_list, "bundle.type_mismatch", "Zip file not accepted")
        return None
    if not is_dir and not is_zip:
        add_issue(issues, path_list, "bundle.invalid", f"Not a directory or zip file: {path}")
        return None

    # Check zip extension
    if is_zip and zip_ext:
        if not path.endswith(f'.{zip_ext}'):
            add_issue(issues, path_list, "bundle.wrong_ext", f"Expected .{zip_ext} extension")
            return None

    # Create context
    try:
        ctx = FSContext(path)
    except Exception as e:
        add_issue(issues, path_list, "bundle.open_error", str(e))
        return None

    # Validate name pattern
    if name_pattern:
        name = ctx.basename()
        if not re.match(name_pattern, name):
            add_issue(issues, path_list, "bundle.name_mismatch", f"Name '{name}' does not match pattern")

    # Validate content
    if content_validator:
        content_validator(ctx, path_list, issues)

    return ctx


def validate_json_file(ctx: FSContext, rel_path: str, path: list[str], issues: Issues,
                       content_validator: Validator | None = None) -> Any | None:
    """Validate a JSON file within a bundle context."""
    file_path = path + [rel_path]

    if not ctx.exists(rel_path):
        add_issue(issues, file_path, "file.not_found", f"File not found: {rel_path}")
        return None

    if not ctx.is_file(rel_path):
        add_issue(issues, file_path, "file.not_file", f"Not a file: {rel_path}")
        return None

    try:
        content = ctx.read_json(rel_path)
    except json.JSONDecodeError as e:
        add_issue(issues, file_path, "json.parse_error", f"Invalid JSON: {e}")
        return None
    except Exception as e:
        add_issue(issues, file_path, "file.read_error", str(e))
        return None

    if content_validator:
        content_validator(content, file_path, issues)

    return content


def validate_fs_file(ctx: FSContext, rel_path: str, path: list[str], issues: Issues,
                     ext: str | None = None) -> bool:
    """Validate a file exists within a bundle context."""
    file_path = path + [rel_path]

    if not ctx.exists(rel_path):
        add_issue(issues, file_path, "file.not_found", f"File not found: {rel_path}")
        return False

    if not ctx.is_file(rel_path):
        add_issue(issues, file_path, "file.not_file", f"Not a file: {rel_path}")
        return False

    if ext:
        actual_ext = rel_path.rsplit('.', 1)[-1] if '.' in rel_path else ''
        if actual_ext != ext:
            add_issue(issues, file_path, "file.wrong_ext", f"Expected .{ext}, got .{actual_ext}")
            return False

    return True


def validate_fs_directory(ctx: FSContext, rel_path: str, path: list[str], issues: Issues) -> bool:
    """Validate a directory exists within a bundle context."""
    dir_path = path + [rel_path]

    if not ctx.exists(rel_path):
        add_issue(issues, dir_path, "dir.not_found", f"Directory not found: {rel_path}")
        return False

    if not ctx.is_dir(rel_path):
        add_issue(issues, dir_path, "dir.not_dir", f"Not a directory: {rel_path}")
        return False

    return True


# === Entry points ===

def validate(value: Any, validator: Validator) -> dict[str, Any]:
    """Run validation on a value and return result."""
    issues: Issues = []
    validator(value, [], issues)
    return {
        "ok": len(issues) == 0,
        "issues": issues
    }


def validate_path(path: str, validator: Callable[[str, list[str], Issues], FSContext | None]) -> dict[str, Any]:
    """Run validation on a file system path and return result."""
    issues: Issues = []
    ctx = validator(path, [], issues)
    if ctx:
        ctx.close()
    return {
        "ok": len(issues) == 0,
        "issues": issues
    }

File: src/codegen/swift/prelude.swift
// SpecSpec Swift Prelude
// Validation primitives - embedded at top of generated validators

import Foundation

// MARK: - Types

public struct Issue: Codable {
    public let path: String
    public let code: String
    public let message: String
}

public typealias Issues = [Issue]
public typealias Validator = (Any, [String], inout Issues) -> Void

public struct ValidationResult: Codable {
    public let ok: Bool
    public let issues: Issues
}

private func addIssue(_ issues: inout Issues, _ path: [String], _ code: String, _ message: String) {
    issues.append(Issue(path: path.isEmpty ? "(root)" : path.joined(separator: "."), code: code, message: message))
}

// MARK: - Primitive Validators

public func validateStr(_ value: Any, _ path: [String], _ issues: inout Issues,
                        minLength: Int? = nil, maxLength: Int? = nil, pattern: String? = nil) {
    guard let str = value as? String else {
        addIssue(&issues, path, "type.mismatch", "Expected string, got \(type(of: value))")
        return
    }
    if let min = minLength, str.count < min {
        addIssue(&issues, path, "str.too_short", "String length \(str.count) is less than minimum \(min)")
    }
    if let max = maxLength, str.count > max {
        addIssue(&issues, path, "str.too_long", "String length \(str.count) exceeds maximum \(max)")
    }
    if let p = pattern {
        let regex = try? NSRegularExpression(pattern: p)
        let range = NSRange(str.startIndex..., in: str)
        if regex?.firstMatch(in: str, range: range) == nil {
            addIssue(&issues, path, "str.pattern_mismatch", "String does not match pattern \(p)")
        }
    }
}

public func validateNum(_ value: Any, _ path: [String], _ issues: inout Issues,
                        min: Double? = nil, max: Double? = nil, integer: Bool = false) {
    let num: Double
    if let n = value as? Double {
        num = n
    } else if let n = value as? Int {
        num = Double(n)
    } else if let n = value as? NSNumber {
        num = n.doubleValue
    } else {
        addIssue(&issues, path, "type.mismatch", "Expected number, got \(type(of: value))")
        return
    }

    if integer && num != num.rounded() {
        addIssue(&issues, path, "num.not_integer", "Expected integer, got \(num)")
    }
    if let m = min, num < m {
        addIssue(&issues, path, "num.too_small", "Number \(num) is less than minimum \(m)")
    }
    if let m = max, num > m {
        addIssue(&issues, path, "num.too_large", "Number \(num) exceeds maximum \(m)")
    }
}

public func validateBool(_ value: Any, _ path: [String], _ issues: inout Issues) {
    if !(value is Bool) {
        addIssue(&issues, path, "type.mismatch", "Expected boolean, got \(type(of: value))")
    }
}

public func validateLiteral<T: Equatable>(_ value: Any, _ path: [String], _ issues: inout Issues, _ expected: T) {
    guard let actual = value as? T else {
        addIssue(&issues, path, "literal.mismatch", "Expected \(expected), got \(value)")
        return
    }
    if actual != expected {
        addIssue(&issues, path, "literal.mismatch", "Expected \(expected), got \(actual)")
    }
}

public func validatePattern(_ value: Any, _ path: [String], _ issues: inout Issues, _ pattern: String) {
    guard let str = value as? String else {
        addIssue(&issues, path, "type.mismatch", "Expected string for pattern match, got \(type(of: value))")
        return
    }
    let regex = try? NSRegularExpression(pattern: pattern)
    let range = NSRange(str.startIndex..., in: str)
    if regex?.firstMatch(in: str, range: range) == nil {
        addIssue(&issues, path, "pattern.mismatch", "Value does not match pattern \(pattern)")
    }
}

// MARK: - Structural Validators

public func validateObject(_ value: Any, _ path: [String], _ issues: inout Issues) -> Bool {
    if value is [String: Any] {
        return true
    }
    addIssue(&issues, path, "type.mismatch", "Expected object, got \(type(of: value))")
    return false
}

public func validateField(_ obj: Any, _ path: [String], _ issues: inout Issues,
                          _ key: String, validator: Validator? = nil, optional: Bool = false) {
    guard let dict = obj as? [String: Any] else { return }

    if dict[key] == nil {
        if !optional {
            addIssue(&issues, path, "field.missing", "Missing required field: \(key)")
        }
        return
    }

    if let v = validator {
        v(dict[key]!, path + [key], &issues)
    }
}

public func validateList(_ value: Any, _ path: [String], _ issues: inout Issues,
                         itemValidator: Validator? = nil, minItems: Int? = nil, maxItems: Int? = nil) {
    guard let arr = value as? [Any] else {
        addIssue(&issues, path, "type.mismatch", "Expected array, got \(type(of: value))")
        return
    }

    if let min = minItems, arr.count < min {
        addIssue(&issues, path, "list.too_short", "Array length \(arr.count) is less than minimum \(min)")
    }
    if let max = maxItems, arr.count > max {
        addIssue(&issues, path, "list.too_long", "Array length \(arr.count) exceeds maximum \(max)")
    }

    if let iv = itemValidator {
        for (i, item) in arr.enumerated() {
            iv(item, path + ["[\(i)]"], &issues)
        }
    }
}

public func validateOneOf(_ value: Any, _ path: [String], _ issues: inout Issues,
                          _ validators: [Validator]) {
    for validator in validators {
        var testIssues: Issues = []
        validator(value, path, &testIssues)
        if testIssues.isEmpty {
            return // Matched
        }
    }
    addIssue(&issues, path, "oneof.no_match", "Value does not match any of the options")
}

// MARK: - File System Context

public class FSContext {
    public let basePath: String
    public let isZip: Bool
    private var zipEntries: [String: Data] = [:]

    public init(_ path: String) throws {
        self.basePath = path
        var isDir: ObjCBool = false

        if FileManager.default.fileExists(atPath: path, isDirectory: &isDir) {
            if isDir.boolValue {
                self.isZip = false
            } else if path.hasSuffix(".zip") || path.hasSuffix(".asks") {
                self.isZip = true
                try loadZip()
            } else {
                throw NSError(domain: "FSContext", code: 1, userInfo: [NSLocalizedDescriptionKey: "Not a directory or zip: \(path)"])
            }
        } else {
            throw NSError(domain: "FSContext", code: 2, userInfo: [NSLocalizedDescriptionKey: "Path not found: \(path)"])
        }
    }

    private func loadZip() throws {
        // Simple zip reading - in production use a proper zip library
        let data = try Data(contentsOf: URL(fileURLWithPath: basePath))
        // For simplicity, we'll just store raw data and handle basic cases
        // Real implementation should use ZIPFoundation or similar
        zipEntries["__raw__"] = data
    }

    public func exists(_ relPath: String) -> Bool {
        if isZip {
            return zipEntries[relPath] != nil || zipEntries.keys.contains { $0.hasPrefix(relPath + "/") }
        }
        return FileManager.default.fileExists(atPath: (basePath as NSString).appendingPathComponent(relPath))
    }

    public func isFile(_ relPath: String) -> Bool {
        if isZip {
            return zipEntries[relPath] != nil
        }
        var isDir: ObjCBool = false
        let fullPath = (basePath as NSString).appendingPathComponent(relPath)
        return FileManager.default.fileExists(atPath: fullPath, isDirectory: &isDir) && !isDir.boolValue
    }

    public func isDir(_ relPath: String) -> Bool {
        if isZip {
            return zipEntries.keys.contains { $0.hasPrefix(relPath + "/") }
        }
        var isDir: ObjCBool = false
        let fullPath = (basePath as NSString).appendingPathComponent(relPath)
        return FileManager.default.fileExists(atPath: fullPath, isDirectory: &isDir) && isDir.boolValue
    }

    public func read(_ relPath: String) throws -> String {
        if isZip {
            guard let data = zipEntries[relPath] else {
                throw NSError(domain: "FSContext", code: 3, userInfo: [NSLocalizedDescriptionKey: "File not found in zip: \(relPath)"])
            }
            return String(data: data, encoding: .utf8) ?? ""
        }
        let fullPath = (basePath as NSString).appendingPathComponent(relPath)
        return try String(contentsOfFile: fullPath, encoding: .utf8)
    }

    public func readJson(_ relPath: String) throws -> Any {
        let content = try read(relPath)
        guard let data = content.data(using: .utf8) else {
            throw NSError(domain: "FSContext", code: 4, userInfo: [NSLocalizedDescriptionKey: "Invalid UTF-8"])
        }
        return try JSONSerialization.jsonObject(with: data)
    }

    public func basename() -> String {
        var name = (basePath as NSString).lastPathComponent
        if let dotIdx = name.lastIndex(of: ".") {
            name = String(name[..<dotIdx])
        }
        return name
    }
}

public typealias FSValidator = (FSContext, [String], inout Issues) -> Void

// MARK: - File System Validators

public func validateBundle(_ bundlePath: String, _ pathList: [String], _ issues: inout Issues,
                           acceptDir: Bool = true, acceptZip: Bool = false,
                           zipExt: String? = nil, namePattern: String? = nil,
                           contentValidator: FSValidator? = nil) -> FSContext? {
    var isDir: ObjCBool = false
    let exists = FileManager.default.fileExists(atPath: bundlePath, isDirectory: &isDir)

    if !exists {
        addIssue(&issues, pathList, "bundle.not_found", "Path not found: \(bundlePath)")
        return nil
    }

    let isDirectory = isDir.boolValue
    let isZipFile = !isDirectory && (bundlePath.hasSuffix(".zip") || (zipExt != nil && bundlePath.hasSuffix(".\(zipExt!)")))

    if isDirectory && !acceptDir {
        addIssue(&issues, pathList, "bundle.type_mismatch", "Directory not accepted")
        return nil
    }
    if isZipFile && !acceptZip {
        addIssue(&issues, pathList, "bundle.type_mismatch", "Zip file not accepted")
        return nil
    }
    if !isDirectory && !isZipFile {
        addIssue(&issues, pathList, "bundle.invalid", "Not a valid bundle: \(bundlePath)")
        return nil
    }

    do {
        let ctx = try FSContext(bundlePath)

        if let pattern = namePattern {
            let name = ctx.basename()
            let regex = try? NSRegularExpression(pattern: pattern)
            let range = NSRange(name.startIndex..., in: name)
            if regex?.firstMatch(in: name, range: range) == nil {
                addIssue(&issues, pathList, "bundle.name_mismatch", "Name '\(name)' does not match pattern")
            }
        }

        if let cv = contentValidator {
            cv(ctx, pathList, &issues)
        }

        return ctx
    } catch {
        addIssue(&issues, pathList, "bundle.open_error", error.localizedDescription)
        return nil
    }
}

public func validateJsonFile(_ ctx: FSContext, _ relPath: String, _ path: [String], _ issues: inout Issues,
                             contentValidator: Validator? = nil) -> Any? {
    let filePath = path + [relPath]

    if !ctx.exists(relPath) {
        addIssue(&issues, filePath, "file.not_found", "File not found: \(relPath)")
        return nil
    }

    if !ctx.isFile(relPath) {
        addIssue(&issues, filePath, "file.not_file", "Not a file: \(relPath)")
        return nil
    }

    do {
        let content = try ctx.readJson(relPath)
        if let cv = contentValidator {
            cv(content, filePath, &issues)
        }
        return content
    } catch {
        addIssue(&issues, filePath, "json.parse_error", "Invalid JSON: \(error.localizedDescription)")
        return nil
    }
}

public func validateFsFile(_ ctx: FSContext, _ relPath: String, _ path: [String], _ issues: inout Issues,
                           ext: String? = nil) -> Bool {
    let filePath = path + [relPath]

    if !ctx.exists(relPath) {
        addIssue(&issues, filePath, "file.not_found", "File not found: \(relPath)")
        return false
    }

    if !ctx.isFile(relPath) {
        addIssue(&issues, filePath, "file.not_file", "Not a file: \(relPath)")
        return false
    }

    if let e = ext {
        let actualExt = (relPath as NSString).pathExtension
        if actualExt != e {
            addIssue(&issues, filePath, "file.wrong_ext", "Expected .\(e), got .\(actualExt)")
            return false
        }
    }

    return true
}

public func validateFsDirectory(_ ctx: FSContext, _ relPath: String, _ path: [String], _ issues: inout Issues) -> Bool {
    let dirPath = path + [relPath]

    if !ctx.exists(relPath) {
        addIssue(&issues, dirPath, "dir.not_found", "Directory not found: \(relPath)")
        return false
    }

    if !ctx.isDir(relPath) {
        addIssue(&issues, dirPath, "dir.not_dir", "Not a directory: \(relPath)")
        return false
    }

    return true
}

// MARK: - Entry Points

public func validate(_ value: Any, _ validator: Validator) -> ValidationResult {
    var issues: Issues = []
    validator(value, [], &issues)
    return ValidationResult(ok: issues.isEmpty, issues: issues)
}

public func validatePath(_ bundlePath: String,
                         _ validator: @escaping (String, [String], inout Issues) -> FSContext?) -> ValidationResult {
    var issues: Issues = []
    _ = validator(bundlePath, [], &issues)
    return ValidationResult(ok: issues.isEmpty, issues: issues)
}
